src/main/java/org/javarosa/xml/TreeElementParser.java

```diff
@@ -55,10 +63,83 @@ public TreeElement parse() throws InvalidStructureException, IOException,
                     break;
                 default:
                     throw new InvalidStructureException(
-                            "Exception while trying to parse an XML Tree, got something other than tags and text", parser);
+                        "Exception while trying to parse an XML Tree, got something other than tags and text", parser);
             }
         }
 
         return element;
     }
+
+    /**
+     * Returns the internal secondary instance with the instance ID created
+     * Used for deserializing
+     * @return The TreeElement node of the internal instance
+     * @throws IOException
+     * @throws InvalidReferenceException
+     * @throws InvalidStructureException
+     * @throws XmlPullParserException
+     * @throws UnfullfilledRequirementsException
+     */
+    public TreeElement parseInternalSecondaryInstance() throws IOException, InvalidReferenceException, InvalidStructureException, XmlPullParserException, UnfullfilledRequirementsException {
+        final int depth = parser.getDepth();
+        //instance node is assumed not to be the first node
+        int foundInstanceIndex = -1;
+        while (parser.getDepth() >= depth) {
+            if(isInstanceNode()){
+                //instance names are in the default namespace
+                String instanceId = parser.getAttributeValue( null, ID_ATTR);
+                if (this.instanceId.equals(instanceId)) {
+                    return new TreeElementParser(parser, foundInstanceIndex, instanceId).parse();
+                } else {
+                    skipSubTree();
+                }
+            }
+        }
+        throw new InvalidReferenceException(String.format("The instance ID %s was not found in the XForm file ", instanceId),instanceId);
+    }
+
+    /**
+     *
+     * Parses the internal instances from an XForm (excluding the main instance)
+     * @return List of #TreeElement representation of the internal
+     * instance nodes.
+     *
+     * @throws IOException There was an error parsing the xml file
+     * @throws InvalidStructureException
+     * @throws XmlPullParserException
+     * @throws UnfullfilledRequirementsException
+     */
+    public List<TreeElement> parseInternalSecondaryInstances() throws InvalidStructureException, XmlPullParserException, UnfullfilledRequirementsException, IOException {
+        /**
+         * This could be implemented using the XPath reference to target specifically
+         * the path of the instance node, but currently, this targets any path where
```

Comment:
Could you please explain this concept in greater detail? Evaluating an XPath expression with existing JavaRosa components requires a tree of `TreeElement`s which is what this is building. So it doesn't seem possible to evaluate an XPath expression in the process of building it.